util/topsql/tracecpu/profile.go

```diff
@@ -0,0 +1,421 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tracecpu
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"runtime/pprof"
+	"strconv"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/google/pprof/profile"
+	"github.com/pingcap/tidb/sessionctx/variable"
+	"github.com/pingcap/tidb/util"
+	"github.com/pingcap/tidb/util/logutil"
+	"go.uber.org/zap"
+)
+
+const (
+	labelSQL        = "sql"
+	labelSQLDigest  = "sql_digest"
+	labelPlanDigest = "plan_digest"
+)
+
+// GlobalTopSQLCPUProfiler is the global SQL stats profiler.
+var GlobalTopSQLCPUProfiler = newTopSQLCPUProfiler()
+
+type topSQLCPUProfiler struct {
+	taskCh chan *profileTask
+
+	mu struct {
+		sync.Mutex
+		ept *exportProfileTask
+	}
+	collector atomic.Value
+}
+
+var (
+	defaultProfileBufSize = 100 * 1024
+	profileBufPool        = sync.Pool{
+		New: func() interface{} {
+			return bytes.NewBuffer(make([]byte, 0, defaultProfileBufSize))
+		},
+	}
+)
+
+// newTopSQLCPUProfiler create a topSQLCPUProfiler.
+func newTopSQLCPUProfiler() *topSQLCPUProfiler {
+	return &topSQLCPUProfiler{
+		taskCh: make(chan *profileTask, 128),
+	}
+}
+
+func (sp *topSQLCPUProfiler) Run() {
+	logutil.BgLogger().Info("cpu profiler started")
+	go sp.startCPUProfileWorker()
+	go sp.startAnalyzeProfileWorker()
+}
+
+func (sp *topSQLCPUProfiler) SetCollector(c TopSQLCollector) {
+	sp.collector.Store(c)
+}
+
+func (sp *topSQLCPUProfiler) GetCollector() TopSQLCollector {
+	c, ok := sp.collector.Load().(TopSQLCollector)
+	if !ok || c == nil {
+		return nil
+	}
+	return c
+}
+
+func (sp *topSQLCPUProfiler) startCPUProfileWorker() {
+	defer util.Recover("top-sql", "profileWorker", nil, false)
+	for {
+		if sp.IsEnabled() {
+			sp.doCPUProfile()
+		} else {
+			time.Sleep(time.Second)
+		}
+	}
+}
+
+func (sp *topSQLCPUProfiler) doCPUProfile() {
+	intervalSecond := variable.TopSQLVariable.PrecisionSeconds.Load()
+	task := sp.newProfileTask()
+	if err := pprof.StartCPUProfile(task.buf); err != nil {
+		// Sleep a while before retry.
+		time.Sleep(time.Second)
+		sp.putTaskToBuffer(task)
+		return
+	}
+	ns := int64(time.Second)*intervalSecond - int64(time.Now().Nanosecond())
+	time.Sleep(time.Nanosecond * time.Duration(ns))
+	pprof.StopCPUProfile()
+	task.end = time.Now().Unix()
+	sp.taskCh <- task
+}
+
+func (sp *topSQLCPUProfiler) startAnalyzeProfileWorker() {
+	defer util.Recover("top-sql", "analyzeProfileWorker", nil, false)
+	for {
+		task := <-sp.taskCh
+		p, err := profile.ParseData(task.buf.Bytes())
+		if err != nil {
+			logutil.BgLogger().Error("parse profile error", zap.Error(err))
+			sp.putTaskToBuffer(task)
+			continue
+		}
+		stats := sp.parseCPUProfileBySQLLabels(p)
+		sp.handleExportProfileTask(p)
+		if c := sp.GetCollector(); c != nil {
+			c.Collect(task.end, stats)
+		}
+		sp.putTaskToBuffer(task)
+	}
+}
+
+type profileTask struct {
```

Comment:
This is not even a task, right? A task is something that can start and finish. You generates a new "task" from the buffer and put the "task" back to buffer, which indicates that the "task" does not have such semantics.

Reply:
em....How about rename to `profileTask`?